crates/bevy_asset/macros/src/lib.rs

```diff
@@ -41,33 +41,101 @@ fn derive_dependency_visitor_internal(
     ast: &DeriveInput,
     bevy_asset_path: &Path,
 ) -> Result<proc_macro2::TokenStream, syn::Error> {
-    let mut field_visitors = Vec::new();
-    if let Data::Struct(data_struct) = &ast.data {
-        for field in &data_struct.fields {
-            if field
-                .attrs
-                .iter()
-                .any(|a| a.path().is_ident(DEPENDENCY_ATTRIBUTE))
-            {
-                if let Some(field_ident) = &field.ident {
-                    field_visitors.push(quote! {
-                        #bevy_asset_path::VisitAssetDependencies::visit_dependencies(&self.#field_ident, visit);
-                    });
+    let struct_name = &ast.ident;
+    let (impl_generics, type_generics, where_clause) = &ast.generics.split_for_impl();
+
+    let body = match &ast.data {
+        Data::Struct(data_struct) => {
+            let mut field_visitors = Vec::new();
+
+            for field in &data_struct.fields {
+                if field
+                    .attrs
+                    .iter()
+                    .any(|a| a.path().is_ident(DEPENDENCY_ATTRIBUTE))
+                {
+                    if let Some(field_ident) = &field.ident {
+                        field_visitors.push(quote! {
+                            #bevy_asset_path::VisitAssetDependencies::visit_dependencies(&self.#field_ident, visit);
+                        });
+                    }
                 }
             }
```

Comment:
No need to accumulate the `TokenStream`s into `Vec`s, the `#(#foo)*` syntax supports any iterator.

something like the following should be enough:

```rust
let field_visitors = data_struct.fields.iter().filter_map(|field| {
    // ...
});

Reply:
i'm happy to change it; this is cart's code though. isn't there some kind of law against that?

Reply:
Well, it would be a good start avoiding to duplicate code :P I probably wouldn't have said anything if you just extracted the code and used a function twice. But the macro code is very crufty. I think it's important for clarity to reduce the boilerplate. (beside, reducing allocations can improve compile times)